### DISCUSSION TITLE
Typescript, schemas em vários lugares, é isso mesmo?

### DISCUSSION THREAD

**Discussion by wilsonneto-swe** [frontendbr/forum#1663] [Perguntas e Respostas]
Fala galera, estou começando a fazer meus primeiros projetos para produção em node.

Reparei que precisa declarar o schema em muitos locais diferentes, e na maior parte das vezes o mesmo esquema... (Usando Express, JOI e Mongoose) exemplo:

- Declarei o schema do usuário no Joi para validar as entradas do meu controller 
- Declarei o mesmo schema no mongoose
- Declarei o mesmo schema como uma interface typescript que meus métodos usam...

É isso  mesmo? Preciso declarar em cada um, como fazem? Existe uma maneira de o JOI e o Mongoose se guiarem pelo tipo/interface do typescript mesmo?

Abraçoos!

**Comment by eliseumds** [comment]
Sim, normalmente você vai precisar de uma interface escrita em duas ou mais linguagens/ferramentas diferentes. Isso não significa que esse processo não possa ser automatizado, por ex: Prisma.io, Refit e Scraml.

**Comment by Sergioamjr** [comment]
Se eles forem iguais, você pode defini-los em um arquivo, exporta-los e usa-los em outros lugares. Fica mais fácil de dar manutenção.

**Comment by wilsonneto-swe** [comment]
@Sergioamjr, diga mais sobre por favor, como eu definiria um arquivo com o Schema do Mongoose, o esquema do JOI e o do TS no mesmo arquivo? Como seria possível, tem alguma idéia?

**Comment by Sergioamjr** [comment]
Claro, @willycamargo Eu havia pensando numa estrutura assim
```
controllers/
models/
types/
    user.js
```

```javascript
//types/user.js
const Joi = require('@hapi/joi');

const joiSchema = Joi.object({
    name: Joi.string().required(),
    email: Joi.string().required(),
    telefone: Joi.string().required(),
    nascimento: Joi.string(),
});

const mongooseSchema = {
    nome: { type: String, required: true },
    email: { type: String, required: true },
    telefone: { type: String, required: true },
    nascimento: { type: String }
};

interface tsSchema {
    name: String;
    email: String;
    telefone: String;
    nascimento: String;
}

module.exports = {
    joiSchema, 
    mongooseSchema,
    tsSchema
}
```

E pra usar em qualquer lugar, seria só importar

```javascript
const { joiSchema, mongooseSchema, tsSchema } = require('../types/user.js')
// ou
const UserTypes = require('../types/user.js')
// UserTypes.joiSchema || UserTypes.mongooseSchema || UserTypes.tsSchema
```

Procurando alguns exemplos na internet, eu encontrei [essa lib](https://github.com/yoitsro/joigoose) que você pode usar o mesmo schema pro joi e pro mongoose.

**Comment by wilsonneto-swe** [comment]
Shoow @Sergioamjr, entendi, verdade. Eu esbarrei nessa lib e estou cogitando usá-la, queria mais pegar uma opinião geral ou ver se é assim mesmo, e bom, é assim mesmo rs Vou adotar esta abordagem mesmo de colocar os schemas no mesmo arquivo... Já nos projetos que eu for separar em camadas ai vou acabar tendo os schemas "duplicados" em vários lugares mesmo. 

Muito obrigado!
